fix(CreateChatView): fixed tagselector various issues

alexandraB99 opened the pull request:
Closes #7340

### What does the PR do
CreateChatView fixed tagselector various issues

### Affected areas
CreateChatView

### StatusQ checklist

- [ ] add documentation if necessary (new component, new feature)
- [ ] update sandbox app
  - in case of new component, add new component page
  - in case of new features, add variation to existing component page
  - nice to have: add it to the demo application as well
- [ ] test changes in both light and dark theme?

### Screenshot of functionality (including design for comparison)


https://user-images.githubusercontent.com/31625338/194599748-cec9ffff-87c2-4b94-a046-5876fe839883.mov

https://user-images.githubusercontent.com/31625338/194599756-1729e3eb-b4e9-4adf-8863-4d8783cf070c.mov

https://user-images.githubusercontent.com/31625338/194599782-61b00742-9b8f-4d8b-82f3-335f1d8c03b3.mov

https://user-images.githubusercontent.com/31625338/194599793-c58aeae5-8607-4e98-b8b4-58a359b1b6bb.mov

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | a3426974 | [#1](https://ci.status.im/job/status-desktop/job/branches/job/linux-cpp/job/PR-7668/1/) | 2022-09-29 15:48:26 | ~5 min | `linux-cpp` | [:package:`pkg`](https://ci.status.im/job/status-desktop/job/branches/job/linux-cpp/job/PR-7668/1//consoleText) |

glitchminer commented:
## Issues found

---
#### 1. OPEN Only 19 users (including admin) can be added to group
Expected:
20 users can be added

---
#### 2. OPEN (OBSERVATION) Can only scroll with trackpad when hovering over scrollbar
When the pointer is over a pill it is not possible to scroll 

---
#### 3. OPEN (OBSERVATION) Scrollbar has been moved within the input field instead of on the edge
This means it now overlaps with the user pills 
**Before**
![2022-09-08_20-02-39 (1)](https://user-images.githubusercontent.com/50769329/189205597-ec370872-b8ba-4713-8d82-107cf81cf235.gif)

**After**
![image](https://user-images.githubusercontent.com/50769329/193348384-dab819a7-a0ea-4986-b2ca-48e4b0453448.png)

---
#### 4. OPEN When 4 users pills are added the left margin increases
![image](https://user-images.githubusercontent.com/50769329/193349177-ce33b32a-ead8-4e2c-a4c8-e41616c5b406.png)

---
#### 5. OPEN Admin pill can be removed with backspace
![image](https://user-images.githubusercontent.com/50769329/193349620-b269fac3-3492-427a-81df-153ab3a3844c.png)

---
#### 6. OPEN Clicking on confirm adds a contact instead of creating group chat
https://user-images.githubusercontent.com/50769329/193350291-fa62e113-79e7-4593-83fc-f4cd626367ce.mp4

---
#### 7. OPEN (OBSERVATION) On resize, scrollbar does not adjust until in focus
https://user-images.githubusercontent.com/50769329/193351065-5c28f7cc-92d2-4d6a-9cd3-dfb37c60e096.mp4

noeliaSD reviewed (commented):
Apart from all issues @glitchminer listed, another issue still remaining: it is not possible to use arrow keys to navigate between members.

alexandraB99 commented:
> Apart from all issues @glitchminer listed, another issue still remaining: it is not possible to use arrow keys to navigate between members.

@noeliaSD I've replied this in the [issue](https://github.com/status-im/status-desktop/issues/7340), this is not a regression, it was never supported. I would see it more like a feature, nevertheless we need to think if we really want to add it because the tags currently are not part of the input thus the cursor cannot just go inside those. It would need bigger effort to be fixed.

borismelnik commented on ui/app/AppLayouts/Chat/panels/InlineSelectorPanel.qml:
```diff
@@ -58,15 +60,30 @@ Item {
+                        }
+                    }
+                    onWidthChanged: {
+                        if ((initWidth !== width) && (scrollView.contentWidth > (parent.width))) {
```
Small suggestion: remove parentheses around `parent.width` :)

borismelnik commented on ui/app/AppLayouts/Chat/panels/InlineSelectorPanel.qml:
```diff
@@ -58,15 +60,30 @@ Item {
                         onContentWidthChanged: {
-                            if (scrollView.contentWidth > scrollView.width) {
-                                scrollView.contentX = scrollView.contentWidth - scrollView.width
+                            if (scrollView.contentWidth > (parent.width/2)) {
```
Small suggestion: Add spaces between members in math operations :)

osmaczko replied:
Why is that? Why it is scrolled just when `> parent.width/2`?

osmaczko commented on ui/app/AppLayouts/Chat/panels/InlineSelectorPanel.qml:
```diff
@@ -58,15 +60,30 @@ Item {
                     Layout.fillWidth: true
                     Layout.fillHeight: true
+                    property real initWidth: 0.0
+                    Connections {
+                        target: Global
+                        function onOpenCreateChatView() {
+                            tagsItemWrapper.initWidth = width;
+                        }
+                    }
```
This seems wrong. Why generic `InlineSelectorPanel` knows anything about `onOpenCreateChatView` signal?

osmaczko commented on ui/app/AppLayouts/Chat/panels/InlineSelectorPanel.qml:
```diff
@@ -58,15 +60,30 @@ Item {
-                            if (scrollView.contentWidth > scrollView.width) {
-                                scrollView.contentX = scrollView.contentWidth - scrollView.width
+                            if (scrollView.contentWidth > (parent.width/2)) {
+                                scrollView.contentX = scrollView.contentWidth - parent.width + (warningLabel.visible ? (parent.width/2) : (parent.width/3));
```
What is `parent.width/3`? Where does the dimension come from? Is it a warning label size? :eyes:

osmaczko commented on ui/app/AppLayouts/Chat/views/CreateChatView.qml:
```diff
@@ -64,12 +73,16 @@ Page {
-                    root.rootStore.createChatInitMessage = chatInput.textInput.text
-                    root.rootStore.createChatFileUrls = chatInput.fileUrls
-                    createChat()
+                    if (contactsList.visible) {
```
This logic is wrong, see why:

https://user-images.githubusercontent.com/33099791/194327912-507dbcdc-88a1-4d4c-82f3-8e5a2d2bdef2.mp4

osmaczko reviewed (changes requested):
It introduces other issues, members selector behaved better before these changes..

1) Extra space after tags which disappears after pressed

https://user-images.githubusercontent.com/33099791/194329266-a349dadf-7f51-43d9-83d4-21e4ad26cafd.mp4

2) Extra space before tags in some random situations

https://user-images.githubusercontent.com/33099791/194329486-8555499b-3623-44b0-90b6-f1a439900100.mp4

3) Confirm button logic broken

https://user-images.githubusercontent.com/33099791/194329610-7332f20d-8e01-476b-9e80-2e7984f3ad50.mp4

alexandraB99 commented:
@glitchminer updated, please have a look :)
@osmaczko same

alexandraB99 commented:
@osmaczko wanna have a second look?

osmaczko commented:
> @osmaczko wanna have a second look?

Done a quick run, there is an issue of disappearing first member (ourselves) when rejected, see:

https://user-images.githubusercontent.com/33099791/196952055-0fb1fbd9-c3cf-4a4a-887c-8195168dcfc1.mp4

@glitchminer could you please have a second look as well?

osmaczko commented on ui/app/AppLayouts/Chat/views/CreateChatView.qml:
```diff
@@ -81,6 +91,22 @@ Page {
+                }
+
+                Component.onCompleted: {
+                    membersSelector.entryAccepted(d.ownContact);
```
This is so hacky.. :eyes: The proper solution would be to extend the model with ourselves rather than artificially accepting entry.

alexandraB99 replied:
so hacky? I don't see any "hack" on adding an entry manually when opening the view, you say to actually add to the contacts model our own self and still call it contacts model? what about the other places that model is used? this will be "hacky" then to have to exclude the first own entry from all other places it's used.

osmaczko replied:
I am not saying to add it to the global contacts model, just to the model used here.

osmaczko replied:
See: https://github.com/status-im/status-desktop/pull/7990

osmaczko commented on ui/app/AppLayouts/Chat/views/CreateChatView.qml:
```diff
@@ -29,6 +29,15 @@ Page {
 
+    QtObject {
+        id: d
+        property var ownContact: ContactListItemDelegate {
```
That's wrong.. :eyes: What if the internal implementation of MembersSelector is changed?

alexandraB99 replied:
wrong? I don't think it's that dramatic - I guess it will be only a matter of updating. do you have any other suggestion?

osmaczko replied:
It is indeed dramatic because we shouldn't make our code tightly coupled, where one innocently looking change breaks other places. As a client, you shouldn't rely on the internals of the `MembersSelector` component. 

This is my suggestion: https://github.com/status-im/status-desktop/pull/7990

alexandraB99 replied:
@osmaczko thanks! updated, please check again

osmaczko reviewed (changes requested):
Sorry for firing one issue at a time. Right now I can't create a group chat (on master it works). Could you please check?

https://user-images.githubusercontent.com/33099791/197557041-fdcfdd8e-d5c2-465f-9a1c-55590c0ebb23.mp4

alexandraB99 commented:
> Sorry for firing one issue at a time. Right now I can't create a group chat (on master it works). Could you please check?
> create-group-chat-2022-10-24_16.53.33.mp4

@osmaczko fixed!

noeliaSD commented:
There is the squish test `suite_messaging/tst_groupChat` which validates the creation of a group chat and `suite_messaging/tst_adminGroupChat` which edits a created group chat. They are disabled in the CI because they rely on the mailserver and are weak so we still need to work in the test environment to enable them there, but I encourage you to try them locally to check if everything is correct. If you need help checking it out, I can help!

noeliaSD commented:
> There is the squish test `suite_messaging/tst_groupChat` which validates the creation of a group chat and `suite_messaging/tst_adminGroupChat` which edits a created group chat. They are disabled in the CI because they rely on the mailserver and are weak so we still need to work in the test environment to enable them there, but I encourage you to try them locally to check if everything is correct. If you need help checking it out, I can help!

Tested locally and all good!!

osmaczko commented:
1) 
https://user-images.githubusercontent.com/33099791/198271083-a975c01b-7798-4402-968b-619b8dfdcced.mp4

2)
https://user-images.githubusercontent.com/33099791/198271438-21674432-52ff-4901-a46a-7fb5eb84aeac.mp4

osmaczko commented on ui/app/AppLayouts/Chat/views/private/MembersSelectorBase.qml:
```diff
@@ -21,7 +21,7 @@ InlineSelectorPanel {
 
     readonly property int membersLimit: 20 // see: https://github.com/status-im/status-mobile/issues/13066
-    property bool limitReached: model.count >= membersLimit
+    property bool limitReached: (model.count === (membersLimit+1))
```
Why "+1"? Members limit should include ourselves. Also `>=` is safer than `===`.

alexandraB99 commented:
@osmaczko what do you mean with 2? what should be confirmed exactly in this case? a chat to ourselves?

osmaczko commented:
> @osmaczko what do you mean with 2? what should be confirmed exactly in this case? a chat to ourselves?

Admin should be able to remove everyone from the group

alexandraB99 commented:
@osmaczko all fixed! :)

status-im-auto commented:
#### :heavy_check_mark: [status-desktop/branches/uitests/PR-7668#9](https://ci.status.im/job/status-desktop/job/branches/job/uitests/job/PR-7668/9/) :small_blue_diamond: ~21 min :small_blue_diamond: 93a47238 :small_blue_diamond: [:package: e2e package](null)

glitchminer commented:
Hey @alexandraB99, the "king pill" for the current user appears to be missing altogether still.

![image](https://user-images.githubusercontent.com/50769329/200684930-923939f5-e7a9-4670-b0ad-e0a7b096f0e6.png)

Build: StatusIm-Desktop-221107-150606-2f28c3-pr7668.dmg

glitchminer reviewed (commented):
### Issues

---
**1. Backspace key cannot be used to remove first pill**
![image](https://user-images.githubusercontent.com/50769329/201086636-36695610-c951-47c5-ade4-a6215f061e58.png)

---
**2. Formatting when typing after typing when max reached**
![image](https://user-images.githubusercontent.com/50769329/201084620-56f1ca64-eb48-4b36-a304-2c05ce62484c.png)

---
**3. Max number of members is now 21 (contacts + self)**

![image](https://user-images.githubusercontent.com/50769329/201086799-6ab027b7-fbd3-4a7c-9b49-e11cb91eb62f.png)

---
**4. Not possible to scroll after adding max contacts then resizing**

https://user-images.githubusercontent.com/50769329/201086117-a664b763-54f9-4826-81f0-06864f617ae3.mp4

---
**5. Formatting when pasting key/url to add new contact**
![image](https://user-images.githubusercontent.com/50769329/201086415-e88056d8-ad18-40e6-b519-fc855c32fd37.png)

---
**6. "King pill" still shown when editing group**
![image](https://user-images.githubusercontent.com/50769329/201087542-cd616ab6-b2fb-4e37-ae92-185f0819957e.png)

alexandraB99 commented:
> ### Issues
> 
> **1. Backspace key cannot be used to remove first pill** ![image](https://user-images.githubusercontent.com/50769329/201086636-36695610-c951-47c5-ade4-a6215f061e58.png)
> 
> **2. Formatting when typing after typing when max reached** ![image](https://user-images.githubusercontent.com/50769329/201084620-56f1ca64-eb48-4b36-a304-2c05ce62484c.png)
> 
> **3. Max number of members is now 21 (contacts + self)**
> 
> ![image](https://user-images.githubusercontent.com/50769329/201086799-6ab027b7-fbd3-4a7c-9b49-e11cb91eb62f.png)
> 
> **4. Not possible to scroll after adding max contacts then resizing**
> 2022-11-10_11-53-47.mp4
> 
> **5. Formatting when pasting key/url to add new contact** ![image](https://user-images.githubusercontent.com/50769329/201086415-e88056d8-ad18-40e6-b519-fc855c32fd37.png)
> 
> **6. "King pill" still shown when editing group** ![image](https://user-images.githubusercontent.com/50769329/201087542-cd616ab6-b2fb-4e37-ae92-185f0819957e.png)

1) fixed
2) fixed
3) fixed
4) can't reproduce
5) fixed
6) this is ok, we still show the king pill when editing the group

glitchminer commented:
1. CLOSED/FIXED.
2. CLOSED/FIXED
3. OPEN 22 members added instead of 20
![image](https://user-images.githubusercontent.com/50769329/201338248-f4ce4296-ca1e-4b78-891e-6e49ec2eaef7.png)
4. OPEN/CANT REPRODUCE
5. CLOSED/FIXED
6. CLOSED/NA
7. OPEN On edit members - resize width smaller when max contacts reached
![image](https://user-images.githubusercontent.com/50769329/201339254-cef936b2-0d83-4c82-8a11-0539fb607487.png)
8. OPEN Flicker when scrolling

https://user-images.githubusercontent.com/50769329/201339460-5d5369e1-652f-4a42-8fcb-86609ecff929.mp4

9. OPEN Cannot confirm members when only 1 contact added
![image](https://user-images.githubusercontent.com/50769329/201339559-2b28c88c-b51e-494c-8c2e-22f6827bcbe3.png)

alexandraB99 commented:
> 1. CLOSED/FIXED.
> 
>     2. CLOSED/FIXED
> 
>     3. OPEN 22 members added instead of 20
>        ![image](https://user-images.githubusercontent.com/50769329/201338248-f4ce4296-ca1e-4b78-891e-6e49ec2eaef7.png)
> 
>     4. OPEN/CANT REPRODUCE
> 
>     5. CLOSED/FIXED
> 
>     6. CLOSED/NA
> 
>     7. OPEN On edit members - resize width smaller when max contacts reached
>        ![image](https://user-images.githubusercontent.com/50769329/201339254-cef936b2-0d83-4c82-8a11-0539fb607487.png)
> 
>     8. OPEN Flicker when scrolling
> 
> 
> 2022-11-11_12-06-24.mp4
> 
>     9. OPEN Cannot confirm members when only 1 contact added
>        ![image](https://user-images.githubusercontent.com/50769329/201339559-2b28c88c-b51e-494c-8c2e-22f6827bcbe3.png)

3) fixed
7) this needs design input as when the members list is opened and we open to edit the members the available space for group chat edit is reduced.
8) fixed
9) fixed

glitchminer commented:
Hi @alexandraB99 , previous issues resolved. Thanks!
1 new issue added (number 10 here: https://github.com/status-im/status-desktop/pull/7668#issuecomment-1311634224). In edit members the behaviour of the message limit has changed and is not capping at 20 members.

glitchminer reviewed (approved):
LGTM!
